Java/Rathna/Main.java

```diff
@@ -3,10 +3,79 @@
 public class Main {
     public static void main(String[] args) {
         Scanner scan = new Scanner(System.in);
-        System.out.println("Please enter just one string: ");
+        System.out.println("Please enter a string: ");
         String inputStr = scan.next();
         System.out.println("Contents of input string: " + inputStr);
-        if(inputStr.length() >= 1 && inputStr.length() <= 104 )
+        IsValidParenthesis(inputStr);
+        scan.close();
+
+
+        int[] unsortedArray = {29, 10, 14, 30, 37, 14, 18};
+        sortUsingBubbleSortTech(unsortedArray);
+        sortUsingSelectionSortTech(unsortedArray);
+        sortUsingInsertionSortTech(unsortedArray);
+    }
+
+    private static void sortUsingInsertionSortTech(int[] unsortedArray) {
+        for(int j = 1; j < unsortedArray.length ; j++) { //Total TC = O(n^2)
+            for(int i = j - 1; i >= 0; i--){ //Total SC = O(1)
+                if(unsortedArray[j] < unsortedArray[i])
```

Comment:
we could optimize the solution by adding `break` statement in the inner for loop. Whenever we reach the element less than current element, we can break the loop.